Invalid class generated when extracting ABI from enum type

### Current Behavior

This is another problem with ASM calculating the number of runtime visible parameter annotations to put into the class file, similar to https://github.com/gradle/gradle/issues/31416.

This time the issue is with enum constructors having implicit `name` and `ordinal` parameters that cannot be annotated, and this shouldn't be listed.


### Expected Behavior

Extracted ABI classes should contain the right number of annotated parameters, and thus be possible to compile against.

### Context (optional)

The problem is caused by this ASM issue: https://gitlab.ow2.org/asm/asm/-/issues/318023.

### Self-contained Reproducer Project

Given an enum like this:

```java
public enum MyEnum {
    VALUE("meta");

    MyEnum(@RuntimeVisible @RuntimeInvisible String metadata) {
        // Constructor
    }
}
```

...currently the generated class causes the following compilation error:

```text
In /Main.java:4: error: cannot access MyEnum
                        System.out.println("Hello " + MyEnum.class.getName());
                                                      ^
  bad class file: /var/folders/mt/w8xt125171b77qvh3q10jhd00000gn/T/spock_can_compile_against_0_temporaryFolder8504945895062116062/api-stubs/MyEnum.class
    bad RuntimeInvisibleParameterAnnotations attribute: MyEnum(java.lang.String)
    Please remove or make sure it appears in the correct subdirectory of the classpath.
In /Main.java:4: error: cannot access java.lang.Byte
                        System.out.println("Hello " + MyEnum.class.getName());
                                                    ^
```

### Gradle version

8.11

### Build scan URL (optional)

_No response_

### Your Environment (optional)

_No response_

--- a/build-logic/java-api-extractor/src/main/java/org/gradle/internal/tools/api/ApiMemberWriter.java
+++ b/build-logic/java-api-extractor/src/main/java/org/gradle/internal/tools/api/ApiMemberWriter.java
@@ -40,7 +40,7 @@ public interface ApiMemberWriter {
 
     void writeClass(ClassMember classMember, Set<MethodMember> methods, Set<FieldMember> fields, Set<InnerClassMember> innerClasses);
 
-    void writeMethod(/* Nullable */ InnerClassMember declaringInnerClass, MethodMember method);
+    void writeMethod(ClassMember classMember, /* Nullable */ InnerClassMember declaringInnerClass, MethodMember method);
 
     void writeClassAnnotations(Set<AnnotationMember> annotationMembers);
 

--- a/build-logic/java-api-extractor/src/main/java/org/gradle/internal/tools/api/impl/JavaApiMemberWriter.java
+++ b/build-logic/java-api-extractor/src/main/java/org/gradle/internal/tools/api/impl/JavaApiMemberWriter.java
@@ -25,8 +25,10 @@ import org.objectweb.asm.MethodVisitor;
 import org.objectweb.asm.ModuleVisitor;
 import org.objectweb.asm.Type;
 
+import java.util.Optional;
 import java.util.Set;
 
+import static org.objectweb.asm.Opcodes.ACC_ENUM;
 import static org.objectweb.asm.Opcodes.ACC_STATIC;
 
 public class JavaApiMemberWriter implements ApiMemberWriter {
@@ -65,7 +67,7 @@ public class JavaApiMemberWriter implements ApiMemberWriter {
             .findFirst()
             .orElse(null);
         for (MethodMember method : methods) {
-            writeMethod(declaringInnerClass, method);
+            writeMethod(classMember, declaringInnerClass, method);
         }
         for (FieldMember field : fields) {
             FieldVisitor fieldVisitor = apiMemberAdapter.visitField(
@@ -81,24 +83,23 @@ public class JavaApiMemberWriter implements ApiMemberWriter {
     }
 
     @Override
-    public void writeMethod(/* Nullable */ InnerClassMember declaringInnerClass, MethodMember method) {
+    public void writeMethod(ClassMember classMember, /* Nullable */ InnerClassMember declaringInnerClass, MethodMember method) {
         MethodVisitor mv = apiMemberAdapter.visitMethod(
             method.getAccess(), method.getName(), method.getTypeDesc(), method.getSignature(),
             method.getExceptions().toArray(new String[0]));
         writeMethodAnnotations(mv, method.getAnnotations());
         writeMethodAnnotations(mv, method.getParameterAnnotations());
 
-        // Without this parameter annotations for non-static inner class constructors would
-        // end up having one more parameter than they should, because the first parameter
-        // (pointing to `this` of the enclosing type) should not be listed.
+        // In some cases ASM generates the wrong number of parameter annotation entries
+        // in the written classfile. This is a workaround to fix that.
         // See https://gitlab.ow2.org/asm/asm/-/issues/318023
-        if (method.getName().equals("<init>")
-            && declaringInnerClass != null
-            && (declaringInnerClass.getAccess() & ACC_STATIC) != ACC_STATIC) {
-            int parameterCount = Type.getArgumentCount(method.getTypeDesc()) - 1;
-            mv.visitAnnotableParameterCount(parameterCount, true);
-            mv.visitAnnotableParameterCount(parameterCount, false);
-        }
+        calculateNonAnnotableParameterCount(classMember, declaringInnerClass, method)
+            .ifPresent(nonAnnotableParameterCount -> {
+                int totalParameterCount = Type.getArgumentCount(method.getTypeDesc());
+                int annotableParameterCount = totalParameterCount - nonAnnotableParameterCount;
+                mv.visitAnnotableParameterCount(annotableParameterCount, true);
+                mv.visitAnnotableParameterCount(annotableParameterCount, false);
+            });
 
         method.getAnnotationDefaultValue().ifPresent(value -> {
             AnnotationVisitor av = mv.visitAnnotationDefault();
@@ -106,6 +107,22 @@ public class JavaApiMemberWriter implements ApiMemberWriter {
             av.visitEnd();
         });
         mv.visitEnd();
+    }
+
+    private static Optional<Integer> calculateNonAnnotableParameterCount(ClassMember classMember, /* Nullable */ InnerClassMember declaringInnerClass, MethodMember method) {
+        if (method.getName().equals("<init>")) {
+            if ((classMember.getAccess() & ACC_ENUM) == ACC_ENUM) {
+                // Enum constructors have an implicit String and int parameter containing
+                // the name and ordinal of the value that is non-annotable.
+                return Optional.of(2);
+            } else if (declaringInnerClass != null
+                && (declaringInnerClass.getAccess() & ACC_STATIC) != ACC_STATIC) {
+                // Non-static inner-class constructors have an implicit, non-annotable parameter
+                // pointing to the enclosing class instance
+                return Optional.of(1);
+            }
+        }
+        return Optional.empty();
     }
 
     @Override

--- a/build-logic/java-api-extractor/src/test/groovy/org/gradle/internal/tools/api/ApiClassExtractorAnnotationsTest.groovy
+++ b/build-logic/java-api-extractor/src/test/groovy/org/gradle/internal/tools/api/ApiClassExtractorAnnotationsTest.groovy
@@ -16,6 +16,8 @@
 
 package org.gradle.internal.tools.api
 
+import spock.lang.Issue
+
 import java.lang.annotation.ElementType
 import java.lang.reflect.Method
 
@@ -379,6 +381,47 @@ class ApiClassExtractorAnnotationsTest extends ApiClassExtractorTestSupport {
         extractedAnnotation.type() == Integer
         extractedMethods["value"].defaultValue == 42
         extractedAnnotation.value() == 42
+    }
+
+    @Issue("https://github.com/gradle/gradle/issues/31436")
+    def "can compile against extracted enum class with annotations on its constructor"() {
+        given:
+        def api = toApi([
+            'MyEnum': '''
+                public enum MyEnum {
+                    VALUE("meta");
+
+                    MyEnum(@RuntimeVisible @RuntimeInvisible String metadata) {
+                        // Constructor
+                    }
+                }
+            ''',
+            'RuntimeVisible': '''
+                @java.lang.annotation.Retention(java.lang.annotation.RetentionPolicy.RUNTIME)
+                public @interface RuntimeVisible {}
+            ''',
+            'RuntimeInvisible': '''
+                @java.lang.annotation.Retention(java.lang.annotation.RetentionPolicy.CLASS)
+                public @interface RuntimeInvisible {}
+            ''',
+        ])
+
+        def apiStubDir = new File(temporaryFolder, "api-stubs")
+        writeClass(api, 'MyEnum', apiStubDir)
+
+        when:
+        def consumer = compileTo(new File(temporaryFolder, 'consumer'), [
+            'Main': '''
+                public class Main {
+                    public static void main(String[] args) {
+                        System.out.println("Hello " + MyEnum.class.getName());
+                    }
+                }
+            '''
+        ], [apiStubDir])
+
+        then:
+        consumer.classes.Main.clazz.name == "Main"
     }
 
     private static Map<String, Method> mapMethods(Method[] methods) {

--- a/build-logic/java-api-extractor/src/test/groovy/org/gradle/internal/tools/api/ApiClassExtractorInnerClassTest.groovy
+++ b/build-logic/java-api-extractor/src/test/groovy/org/gradle/internal/tools/api/ApiClassExtractorInnerClassTest.groovy
@@ -189,9 +189,8 @@ class ApiClassExtractorInnerClassTest extends ApiClassExtractorTestSupport {
         ])
 
         def apiStubDir = new File(temporaryFolder, "api-stubs")
-        apiStubDir.mkdirs()
         ['Outer', 'Outer$NonStaticInner', 'Outer$StaticInner'].each {
-            new File(apiStubDir, "${it}.class").bytes = api.extractApiClassFrom(api.classes[it])
+            writeClass(api, it, apiStubDir)
         }
 
         when:

--- a/build-logic/java-api-extractor/src/test/groovy/org/gradle/internal/tools/api/ApiClassExtractorTestSupport.groovy
+++ b/build-logic/java-api-extractor/src/test/groovy/org/gradle/internal/tools/api/ApiClassExtractorTestSupport.groovy
@@ -176,6 +176,14 @@ class ApiClassExtractorTestSupport extends Specification {
         toApi([(fqn): script]).classes[fqn]
     }
 
+    @CompileStatic
+    protected static File writeClass(ClassContainer container, name, File targetDir) {
+        targetDir.mkdirs()
+        def classFile = new File(targetDir, "${name}.class")
+        classFile.bytes = container.extractApiClassFrom(container.classes[name])
+        return classFile
+    }
+
     protected void noSuchMethod(Class c, String name, Class... argTypes) {
         try {
             c.getDeclaredMethod(name, argTypes)

--- a/platforms/core-configuration/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/normalization/KotlinApiMemberWriter.kt
+++ b/platforms/core-configuration/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/normalization/KotlinApiMemberWriter.kt
@@ -81,11 +81,15 @@ class KotlinApiMemberWriter private constructor(apiMemberAdapter: ClassVisitor) 
         super.writeClassAnnotations(annotationMembers.filter { it.name != kotlinMetadataAnnotationSignature }.toSet())
     }
 
-    override fun writeMethod(declaringInnerClass: InnerClassMember?, method: MethodMember) {
+    override fun writeMethod(
+        classMember: ClassMember,
+        declaringInnerClass: InnerClassMember?,
+        method: MethodMember
+    ) {
         when {
             method.isInternal() -> return
             method.isInline() -> throw publicInlineFunction(method)
-            else -> super.writeMethod(declaringInnerClass, method)
+            else -> super.writeMethod(classMember, declaringInnerClass, method)
         }
     }
 

